### PR TITLE
fix(ci-analytics): measure default-branch red time per HEAD commit

### PR DESCRIPTION
**Pull request by davincios**
Fixes #

#### Describe the changes you have made in this PR -

`analyze_github_ci_reliability` measured default-branch red time by tracking each workflow as an independent red/green timeline and unioning the periods. GitHub's notion of "the branch is red" is different: it is the check state of the branch's **latest commit**. The two models diverge whenever a workflow does not run on the next commit (path filter, conditional, rename, deletion) — the old model kept the branch red until that workflow happened to run and pass again, while GitHub already showed green. It also hid *which* check made the branch red, so a failing dependency-audit job read as "main was broken for 23 days".

- `find_outages` now follows commits: a commit with a failing check opens an outage, the next fully green commit closes it at its last completion, and a commit whose only runs were cancelled decides nothing. `Outage.workflow` became `Outage.workflows` (every workflow red during the period).
- New `workflow_red_hours` attributes the branch red time per workflow, clipped to the branch outages so a share can never exceed the branch figure. An outage counts toward a workflow only when one of its failing runs completed during it, so a stale failure is never blamed for a later outage another workflow caused. It fills the new `WorkflowSummary.red_hours`, surfaces in the tool payload, and the key-results row now reads e.g. `23.1d of 30 days (77.0%), 2 breakages; Security 23.1d, CI 22.4h` when more than one workflow carried red time.
- A same-day snapshot written by an older version (pre-rename outage shape) is treated as absent by `report_text_from_snapshot` instead of raising `TypeError`.
- Shipped benchmarks switched to `langchain-ai/langchain` and `anomalyco/opencode` (docs, tool description, tests, and figures), measured live on 2026-09-11 under the new per-commit model and verified by an independent re-measurement (langchain: 4,520 runs; opencode: 20,624 runs).
- Docs (`docs/cicd-analytics-demo.mdx`, skill `references/metrics.md`) updated to the per-commit definition.

Known limitations (unchanged data-collection trade-offs, follow-up candidates):
- A failed run on main that was re-run to success is listed by GitHub as its latest attempt only, so that red period is undercounted; fixing it requires attempt-history fetches for default-branch runs, which the collector skips for cost.
- A commit that triggers **no** workflow at all (every workflow path-filtered out) is invisible to the collector, which only reads workflow runs; such a commit can neither close nor extend an outage. Seeing it would require fetching the branch commit list — a new API surface for a rare shape (a repo whose every workflow is path-filtered away right after a red commit).

### Demo/Screenshot for feature changes and bug fixes -

Live run against `Tracer-Cloud/opensre-webapp` (30 days) after the change:

```
red_hours=554.50h (23.10d), outages=2, mean_recovery=11.55d
  outage: ('Security',) 08-13 10:09 -> 09-01 15:19
  outage: ('CI', 'Security') 09-03 15:31 -> 09-07 12:52
  red share: Security: 23.10d
  red share: CI: 0.93d
key results row: ('main branch red', '23.1d of 30 days (77.0%), 2 breakages; Security 23.1d, CI 22.4h')
```

Verified by hand against the GitHub UI: every HEAD commit of `main` in those windows carried a failing check run, and the two September red periods are one continuous branch outage (the branch never went fully green between them).

### Follow-up commits on this branch (skill formulation + plan evidence)

Later commits on this branch (`improved skill formulation`, `skill improvements`, `first set of improvements`, and the final push) extend the scope beyond the red-time fix:

- **`cicd-analytics-demo` skill rewritten as a model-followed plan.** The card no longer narrows `tools:`, lists `references:`, or declares `after_tool` menus. The model calls `ask_user_choice` itself for the repository pick (step 2) and the next-step menu (step 5: `Schedule local loops` / `Slack setup` / `Finish`), and each non-`Finish` branch hands off to its sibling skill via `skill_view`. The shell-side "ask the repository before the model runs" path (`demo_repository.py`) therefore applies only to `cicd-reliability-agent`, which keeps its hook. Tests in `test_demo_picker.py`, `test_demo_repository.py`, `test_choice_prompt_demo_repository.py`, and `test_skills_demo.py` pin this contract. The deprecated `v1.4` card copy is removed.
- **Shared skill rules moved to `skills/common/`** (`ask_once.md`, `numbers_from_tools.md`) so any skill can reference them; the loader already resolves `references:` against the skills root.
- **`analyze_github_ci_reliability` returns `response_text` instead of painting the report**; `cicd-reliability-agent` outputs that text once. Render/tool tests updated accordingly.
- **Task-plan completion evidence** (`core/agent_harness/task_plan/evidence.py`, `turns/plan_evidence_hook.py`): the action turn counts successful non-bookkeeping tool returns; an `update_plan` write that newly marks a step `completed` needs at least one such return since the previous write, otherwise the write is refused with a reason the model can act on. Wired through `update_plan_policy`, `action_driver`, the `update_plan` tool, the SPI, and the turn decision trace.
- Small `integrations/github/__init__.py` export changes to match.

Focused local checks for these commits: `make lint`, `make format-check`, `make typecheck`, `check_imports.py --strict`, `lint-imports`, and `uv run pytest -n auto tests/core tests/tools tests/surfaces tests/integrations tests/cli --ignore=tests/cli/test_smoke.py` (10,148 passed, 50 skipped, 2 xfailed).

---

## Code Understanding and AI Usage

**Did you use AI assistance (ChatGPT, Claude, Copilot, etc.) to write any part of this code?**
- [ ] No, I wrote all the code myself
- [x] Yes, I used AI assistance (continue below)

**If you used AI assistance:**
- [x] I have reviewed every single line of the AI-generated code
- [x] I can explain the purpose and logic of each function/component I added
- [x] I have tested edge cases and understand how the code handles them
- [x] I have modified the AI output to follow this project's coding standards and conventions

**Explain your implementation approach:**

The report claimed main was red for 23.1 of 30 days, which read as wrong because the build itself was green; the cause was the per-workflow red-time model plus missing attribution. Alternatives considered: reading `/commits/{sha}/check-runs` per HEAD commit (GitHub-exact, covers re-runs and non-Actions checks, but adds one API call per commit and a second data path) versus recomputing branch state from the workflow runs already collected (chosen: no extra API cost, same data feeds all KPIs). Key pieces: `find_outages` is a small state machine over commits ordered by first queue time; `workflow_red_hours` intersects each workflow's own red spans with the branch outages it actually failed during; `_red_breakdown` in render appends the attribution only when at least two workflows carry a share.

---

## Checklist before requesting a review
- [x] I have added proper PR title and linked to the issue
- [x] I have performed a self-review of my code
- [x] **I can explain the purpose of every function, class, and logic block I added**
- [x] I understand why my changes work and have tested them thoroughly
- [x] I have considered potential edge cases and how my code handles them
- [x] If it is a core feature, I have added thorough tests
- [x] My code follows the project's style guidelines and conventions

---

Focused local checks run: `make lint`, `make format-check`, `make typecheck`, and `uv run pytest tests/tools/test_github_ci_analytics.py tests/integrations/github/test_ci_analytics_snapshots.py tests/tools/test_ci_reliability_loop.py` (84 passed) plus the `integrations/github/tools/` scope-rule targets and `tests/core/agent_harness/prompts/test_skill_metadata.py` (213 passed).

### CODE REVIEW

File: integrations/github/tools/ci_analytics/metrics.py [hunk 1]
```diff
@@ -359,25 +360,92 @@ def classify_failures(
...
+    never runs again (path filter, rename, deletion) cannot keep the branch
+    red once a later commit is fully green. A commit whose counted runs all
+    passed closes the outage at its last completion; a commit with no counted
+    runs (every run cancelled) decides nothing.
+    """
+    counted = [run for run in runs if run.failed or run.succeeded]
+    by_commit: dict[str, list[WorkflowRun]] = defaultdict(list)
+    for run in counted:
+        by_commit[run.head_sha].append(run)
+    commits = sorted(by_commit.values(), key=lambda commit: min(r.created_at for r in commit))
```
[the remark this answers is not in the document]
Comment (davincios) [reply integrations/github/tools/ci_analytics/metrics.py:377]: Valid observation, but this is a data-collection boundary rather than a logic bug: the collector reads workflow runs only, so a commit that triggers no workflow at all produces no row anywhere and is invisible to `find_outages` — there is nothing to close the outage *with*. Seeing runless commits would require fetching the branch commit list, a new API surface for a rare shape (every workflow path-filtered away immediately after a red commit). The common path-filter case — the next commit still runs at least one workflow — is handled and tested (`test_a_workflow_that_skips_the_next_commit_cannot_keep_the_branch_red`). Documented as a known limitation in the PR description alongside the existing re-run-attempt limitation; both share the same cost trade-off and are follow-up candidates together.

File: integrations/github/tools/ci_analytics/metrics.py [hunk 2]
```diff
@@ -359,25 +360,92 @@ def classify_failures(
...
-                )
+                spans.append((open_since, run.completed_at))
                 open_since = None
         if open_since is not None:
-            outages.append(Outage(name, open_since.completed_at, None, open_since.url))
-    return sorted(outages, key=lambda o: o.started_at)
+            spans.append((open_since, now))
+        clipped = _intersect_hours(spans, branch_spans)
+        if clipped > 0:
+            shares[key] = clipped
```
[the remark this answers is not in the document]
Comment (davincios) [reply integrations/github/tools/ci_analytics/metrics.py]: Real bug — fixed in 9b8b99006. `workflow_red_hours` now counts an outage toward a workflow only when one of that workflow's failing runs completed during that outage, so a stale failure whose workflow never ran again is not blamed for a later outage another workflow caused. Regression test: `test_a_stale_failure_is_not_blamed_for_a_later_unrelated_outage`.

**Comment by davincios** [comment]
@greptile review

**Comment by davincios** [comment]
@greptile review

File: integrations/github/tools/ci_analytics/metrics.py [hunk 3]
```diff
@@ -359,25 +360,112 @@ def classify_failures(
...
+    open_workflows: list[str] = []
+    for commit_runs in commits:
+        failing = sorted((run for run in commit_runs if run.failed), key=lambda r: r.completed_at)
+        if failing:
+            if open_since is None:
+                open_since = failing[0].completed_at
+                open_url = failing[0].url
+            for run in failing:
+                if run.workflow not in open_workflows:
+                    open_workflows.append(run.workflow)
```
[the remark this answers is not in the document]
Comment (davincios) [reply integrations/github/tools/ci_analytics/metrics.py]: Fixed in c7800c1f7 — `open_workflows` is now an insertion-ordered `dict[str, None]` with `setdefault`, keeping first-seen order for the outage label with O(1) membership.

**Comment by davincios** [comment]
@greptile review

**Comment by davincios** [comment]
@greptile review

File: core/agent_harness/task_plan/update_plan_policy.py [hunk 4]
```diff
@@ -6,7 +6,63 @@
...
+    ``pending`` straight to ``completed`` is reset regardless — it was never
+    being worked. Any other new completion (from ``in_progress``, a new or
+    renamed step, or a plan written after the work) needs ``evidence``: a
+    non-bookkeeping tool returned since the previous write. Reset steps are
+    returned so the tool result can name them. A write that completes every
+    step is left alone: closing the checklist after the work is not a false
+    tick, and a text-only final step has no tool to show for itself.
+    """
+    if not plan.steps or plan.all_completed:
+        return plan, ()
```
[the remark this answers is not in the document]
Comment (davincios) [reply core/agent_harness/task_plan/update_plan_policy.py]: Valid — fixed in 30f96f618. The `all_completed` early return is gone; `demote_unevidenced_completions` now exempts only the step that was `in_progress` on the stored plan (the text-only closing step), and a plan with no stored prior or a `pending → completed` jump is demoted like any other unevidenced tick. Pinned by `test_a_checklist_cannot_be_born_or_bulk_ticked_complete_without_evidence` (policy) and `test_update_plan_does_not_accept_a_plan_born_complete_before_any_work` (tool).

File: core/agent_harness/prompts/skills/onboarding-github-ci/a-analyzing-github-ci-performance/SKILL.md [hunk 5]
```diff
@@ -1,167 +1,151 @@
...
+### 5. Offer the next step
+
+Call `ask_user_choice` with the title
+`What would you like to do next?` and these options:
+
+- Schedule local loops
+- Slack setup
+- Finish
+
+Complete when the `ask_user_choice` call for this menu has returned in
```
[the remark this answers is not in the document]
Comment (davincios) [reply core/agent_harness/prompts/skills/onboarding-github-ci/a-analyzing-github-ci-performance/SKILL.md:143]: Assessment: the sequencing concern is real, but it is a host rendering gap rather than a plan-gate deadlock. `task_plan_blocks_conclusion` rejects the report-only reply while step 5 is pending and nudges the model on in the same turn, so the menu does open without another user message. What is lost is the report text itself: a no-tool reply is skipped by `ActionRenderObserver._render_intermediate_message` (it expects the driver to stream it as `final_text`), but a rejected conclusion never becomes `final_text`, and `TurnEndEvent` is not mapped by `tuple_payload_from_event`, so the shell never sees it. Text that accompanies a tool call (`has_tool_calls=True`) is rendered live as a note block, so the robust shape is for the report and the `ask_user_choice` call to travel in one message. The skill body is frozen for this PR at the author's request, and surfacing every rejected conclusion host-side would also show premature "done" summaries, so this is left open as a follow-up decision for the author rather than patched here.
